src/OrleansRuntime/Async/CancellationSourcesExtension.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System;
+using System.Threading.Tasks;
+using Orleans.CodeGeneration;
+using Orleans.Runtime;
+using Orleans.Runtime.Providers;
+
+namespace Orleans.Async
+{
+    /// <summary>
+    /// Contains list of cancellation token source corresponding to the tokens
+    /// passed to the related grain activation.
+    /// </summary>
+    internal class CancellationSourcesExtension : ICancellationSourcesExtension
+    {
+        private readonly Lazy<TraceLogger> _logger = new Lazy<TraceLogger>(() =>
+            TraceLogger.GetLogger("CancellationSourcesExtension", TraceLogger.LoggerType.Application));
+        private readonly Interner<Guid, GrainCancellationToken> _cancellationTokens;
+        private static readonly TimeSpan _cleanupFrequency = TimeSpan.FromMinutes(7);
+        private static readonly int _defaultInternerCollectionSize = 31;
+
+        public CancellationSourcesExtension()
+        {
+            _cancellationTokens = new Interner<Guid, GrainCancellationToken>(
+                _defaultInternerCollectionSize,
+                _cleanupFrequency);
+        }
+
+        public Task CancelTokenSource(GrainCancellationToken token)
+        {
+            GrainCancellationToken gct;
+            if (!_cancellationTokens.TryFind(token.Id, out gct))
+            {
+                _logger.Value.Error(ErrorCode.CancellationTokenCancelFailed,  $"Remote token cancellation failed: token with id {token.Id} was not found");
+                return TaskDone.Done;
+            }
+
+            return gct.Cancel();
+        }
+
+        internal GrainCancellationToken GetOrCreateCancellationToken(GrainCancellationToken token)
```

Comment:
but in that case, why wouldn't the caller just use the GrainCancellationToken that they currently have a reference to, if they don't care about it being interned? I added the comment because I looked at the API and at first didn't understand exactly why the caller even needed to call this method